File: python/medium/memoize.py
def memoize(fn):
    cache = {}
    def inner(*args):
        if args in cache:
            return cache[args]
        result = fn(*args)
        cache[args] = result
        return result
    return inner

counter = 0

def traced_sum(*args):
    global counter
    counter += 1
    return sum(args)

memoized_fn = memoize(traced_sum)

memoized_fn(2,3)
memoized_fn(2,3)

print(counter)

File: ts/medium/memoize.ts
function memoize(fn: Function): Function {
  const cache: Record<string, string> = {};
  
  return function(...args: any) {
    const candidateKey = JSON.stringify(args);
    if (cache.hasOwnProperty(candidateKey)) {
      return cache[candidateKey];
    } else {
      cache[candidateKey] = fn(...args);
      return cache[candidateKey];
    }
  }
}

let callCount = 0;

const memoizedFn = memoize(function (a: any, b: any) {
  callCount += 1;
  return a + b;
});
memoizedFn(2, 3) // 5
memoizedFn(2, 3) // 5
console.log(callCount) // 1 